### PR TITLE
[pixart_rf]Add parsing report id flow

### PR DESCRIPTION
**Pull request by jyu99**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/pixart-rf/fu-pxi-ble-device.c [hunk 1]
```diff
@@ -210,9 +216,23 @@ fu_pxi_ble_device_search_hid_usage_page(guint8 *report_descriptor,
...
+				if(report_descriptor[index] == 0x85) {
+					index = index + 1;
+					*report_id = report_descriptor[index];
+					g_debug("repor_id: %x  ", *report_id);
+					break;
+				}
+
+				index++;
+			}
+
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-ble-device.c:235]: are you trying to go from usage page to report ID? I merged some code for this very recently: https://github.com/fwupd/fwupd/commit/fca53b31d25a9570de4d5c279e4a023ca1dd42ab

File: plugins/pixart-rf/fu-pxi-ble-device.c [hunk 2]
```diff
@@ -46,6 +49,8 @@ struct _FuPxiBleDevice {
 	FuUdevDevice parent_instance;
 	struct ota_fw_state fwstate;
 	guint8 retransmit_id;
+	guint8 feature_report_id;
+	guint8 input_report_id;
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-ble-device.c:53]: these need to be in the to_string() output too

File: plugins/pixart-rf/fu-pxi-ble-device.c (hunk 1 above)
Comment (jyu99) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: It's useful!!  I replace my code with it

Comment (hughsie) [reply plugins/pixart-rf/fu-pxi-ble-device.c:233]: Yell if you need any sample code.

Comment (jyu99) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: After meet with our OEM customer, they decided to use this version. In the next step, we will integrate the new official HID related API into our plugin

Comment (jyu99) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: Thanks your advising.

Comment (hughsie) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: > After meet with our OEM customer, they decided to use this version

Right, but your OEM customer isn't the one maintaining this code for the next 10 years. I'd like you to use the FuHidDescriptor code before I will approve this PR.

Comment (hughsie) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: Sure! The FuHidDescriptor[1] object subclasess FuFirmware so it's a few lines of code: 

	g_autoptr(GBytes) fw = g_bytes_new(pixart_buffer, pixart_sizeofbuffer);
	g_autoptr(FuHidDescriptor) descriptor = fu_hid_descriptor_new();
	g_autoptr(FuHidReport) report = NULL;
	g_autoptr(FuFirmware) item_id = NULL;

	if (!fu_firmware_parse(descriptor, fw, FWUPD_INSTALL_FLAG_NONE, error))
		return FALSE;
	report = fu_hid_descriptor_find_report_by_usage(FU_HID_DESCRIPTOR(descriptor),
							pixart_usage_page, 
							pixart_usage,
							error);
	if (report == NULL)
		return FALSE;
	item_id = fu_firmware_get_image_by_id(FU_FIRMWARE(report), "report-id", error);
	if (item_id == NULL)
		return FALSE;
	pixart_report_id = fu_hid_report_item_get_value(FU_HID_REPORT_ITEM(item_id));

[1] https://github.com/fwupd/fwupd/blob/main/libfwupdplugin/fu-hid-descriptor.h

Comment (jyu99) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: 0x06, 0x02, 0xFF,  // Usage Page (Vendor Defined 0xFF02)
    0x09, 0x01,        // Usage (0x01)
    0xA1, 0x01,        // Collection (Application)
    0x75, 0x08,        //   Report Size (8)
    0x95, 0x14,        //   Report Count (20)
    0x85, 0x09,        //   Report ID (9)
    0x09, 0x01,        //   Usage (0x01)
    0xB1, 0x02,        //   Feature (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position,Non-volatile)
    0xC0,              // End Collection

Comment (jyu99) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: guint8 pixart_buffer[512] = {0x0};
	g_autoptr(GBytes) fw = g_bytes_new(rpt_desc.value, rpt_desc.size);
	g_autoptr(FuHidDescriptor) descriptor = fu_hid_descriptor_new();
	g_autoptr(FuHidReport) report = NULL;
	g_autoptr(FuFirmware) item_id = NULL;

	if (!fu_firmware_parse(descriptor, fw, FWUPD_INSTALL_FLAG_NONE, error))
		return FALSE;
		
	report = fu_hid_descriptor_find_report_by_usage(FU_HID_DESCRIPTOR(descriptor),
							0xff02, 
							0x01,
							error);
	
	if (report == NULL)
		return FALSE;
	item_id = fu_firmware_get_image_by_id(FU_FIRMWARE(report), "report-id", error);
	if (item_id == NULL)
		return FALSE;
	guint8 pixart_report_id = fu_hid_report_item_get_value(FU_HID_REPORT_ITEM(item_id));
	g_debug("#pixart_report_id: %x\r\n",pixart_report_id);

Comment (jyu99) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: the parsing result the report id is 8, but the correct id is 9, Is any condition I miss ?

Comment (jyu99) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: <pre><font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">#pixart_report_id: 8</font>

<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">target usage_page:01 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit item: 6  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">usage_page:     01 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit pos 285</font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">repor_id: 8  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">target usage_page:02 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit item: 6  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">usage_page:     02 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit pos 303</font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">repor_id: 9  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">target usage_page:00 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">usage_page:     00 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit pos 206</font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">repor_id: 5  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">usage_page:     00 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit pos 257</font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">repor_id: 7  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">report_type: 2  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">SetFeature:     08 28 </font>
<font color="#4E9A06">09:22:18.477 FuPluginPixartRf     </font><font color="#3465A4">SetFeature:     09 10 </font>
<font color="#4E9A06">09:22:18.491 FuPluginPixartRf     </font><font color="#3465A4">SetFeature:     09 23 </font>
<font color="#4E9A06">09:22:18.529 FuPluginPixartRf     </font><font color="#3465A4">GetFeature:     0e 09 23 00 33 2e 30 2e 32 84 d8 </font>
<font color="#4E9A06">09:22:18.529 FuPluginPixartRf     </font><font color="#3465A4">doing fixup for old bluez version</font>
<font color="#4E9A06">09:22:18.529 FuPluginPixartRf     </font><font color="#3465A4">SetFeature:     09 2b </font>
<font color="#4E9A06">09:22:18.574 FuPluginPixartRf     </font><font color="#3465A4">GetFeature:</font>
</pre>

Comment (hughsie) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: one sec; you've found a bug in the descriptor parser. I've fixed the bug, but give me an hour to write some tests.

	@@ -260,6 +265,11 @@ fu_hid_descriptor_find_report_by_usage(FuHidDescriptor *self,
					continue;
			}
	 
	+               /* don't return the top-level collection */
	+               item_id = fu_firmware_get_image_by_idx(FU_FIRMWARE(report), FU_HID_ITEM_TAG_COLLECTION, NULL);
	+               if (item_id != NULL)
	+                       continue;
	+

Comment (hughsie) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: Okay, lets fix the API then :) Give me a few mins.

Comment (hughsie) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: @jyu99 does something like this work? https://github.com/fwupd/fwupd/pull/6015 -- you can then do something like:

	report = fu_hid_descriptor_find_report(FU_HID_DESCRIPTOR(firmware),
					       &error,
					       "usage-page",
					       0xFF02,
					       "usage",
					       0x01,
					       "feature",
					       0x02,
					       NULL);

Comment (jyu99) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: It's fine.

Comment (jyu99) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: please tell me if you complete the HID API fix,and then I will recreate the pull request and update my code.Thanks!

Comment (hughsie) [reply plugins/pixart-rf/fu-pxi-ble-device.c:235]: Okay, merged -- please cherry pick that commit into your branch or rebase on top of master.

File: plugins/pixart-rf/fu-pxi-ble-device.c [hunk 3]
```diff
@@ -185,49 +190,108 @@ fu_pxi_ble_device_get_feature(FuPxiBleDevice *self, guint8 *buf, guint bufsz, GE
...
-	}
+	if (item_id == NULL)
+		return FALSE;
+
+	/* find report-id value */
+	*report_id = fu_hid_report_item_get_value(FU_HID_REPORT_ITEM(item_id));
+
+	g_debug("usage-page: %x report_id:%d",usage_page, *report_id);
 
 	return FALSE; /* finished processing */
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-ble-device.c:235]: nah, if you return FALSE you need to set GError. I think you want to return TRUE here.

File: plugins/pixart-rf/fu-pxi-ble-device.c [hunk 4]
```diff
@@ -185,49 +190,108 @@ fu_pxi_ble_device_get_feature(FuPxiBleDevice *self, guint8 *buf, guint bufsz, GE
...
+					       0x01,
+					       "input",
+					       0x02,
+					       NULL);
+	if (report == NULL)
+		return FALSE;
+
+	/* find report-id */
+	item_id = fu_firmware_get_image_by_id(FU_FIRMWARE(report), "report-id", error);
+
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-ble-device.c:273]: no need for this newline

File: plugins/pixart-rf/fu-pxi-ble-device.c [hunk 5]
```diff
@@ -185,49 +190,108 @@ fu_pxi_ble_device_get_feature(FuPxiBleDevice *self, guint8 *buf, guint bufsz, GE
...
+	if (item_id == NULL)
+		return FALSE;
+
+	/* find report-id value */
+	*report_id = fu_hid_report_item_get_value(FU_HID_REPORT_ITEM(item_id));
+
+	g_debug("usage-page: %x report_id:%d",usage_page, *report_id);
+
+	return FALSE; /* finished processing */
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-ble-device.c:282]: same here, return TRUE for success.

File: plugins/pixart-rf/fu-pxi-ble-device.c [hunk 6]
```diff
@@ -185,49 +190,108 @@ fu_pxi_ble_device_get_feature(FuPxiBleDevice *self, guint8 *buf, guint bufsz, GE
...
 
 static gboolean
 fu_pxi_ble_device_check_support_report_id(FuPxiBleDevice *self, GError **error)
 {
 #ifdef HAVE_HIDRAW_H
 	gint desc_size = 0;
+	guint8 report_id = 0;
 	g_autoptr(GByteArray) req = g_byte_array_new();
 
+
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-ble-device.c:294]: no need for two lines

File: plugins/pixart-rf/fu-pxi-ble-device.c [hunk 7]
```diff
@@ -240,25 +304,57 @@ fu_pxi_ble_device_check_support_report_id(FuPxiBleDevice *self, GError **error)
...
+					rpt_desc.size,
+					PXI_HID_DEV_OTA_RETRANSMIT_USAGE_PAGE,
+					&report_id,
+					error)) {
+		self->retransmit_id = report_id;
+	} else {
+
+		self->retransmit_id = PXI_HID_DEV_OTA_RETRANSMIT_REPORT_ID;
 	}
+
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-ble-device.c:331]: the error handling here is funky. I'd have something like `g_autoptr(GError) error_local = NULL;` and then use `&error_local` in `fu_pxi_ble_device_search_hid_feature_report_id()` -- and on failure, do `g_debug("failed to parse descriptor: %s", error_local->message);` and fallback to `self->retransmit_id = PXI_HID_DEV_OTA_RETRANSMIT_REPORT_ID;`

File: plugins/pixart-rf/fu-pxi-ble-device.c [hunk 8]
```diff
@@ -787,6 +883,7 @@ fu_pxi_ble_device_get_model_info(FuPxiBleDevice *self, GError **error)
 			    sizeof(model_name),
 			    error))
 		return FALSE;
+	g_debug("model name%s",model_name);
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-ble-device.c:886]: not sure this debugging is useful -- if it is then it should be in the `_to_string()` output too.

**Comment by hughsie** [comment]
@jyu99 counterproposal in https://github.com/fwupd/fwupd/pull/6027

**Comment by hughsie** [comment]
Merged https://github.com/fwupd/fwupd/pull/6027 instead.

**Pull request closed without merging**